[InferAS] Infer the address space of inttoptr

LuoYuanke opened the pull request:
Currently the InferAddressSpaces would check if the bit value doesn't
change for <ptrtoint, inttoptr> address space cast. However the
condition is too strict. Sometime only the low bit address changes for
swizzling, and the address space is not changed. Take below code as
an example, we can tranform `%gp2 = inttoptr i64 %b to ptr` to
`%gp2 = inttoptr i64 %b to ptr addrspace(2)` and specify addrspace(2)
for the following store instruction.
```
   %gp = addrspacecast ptr addrspace(2) %sp to ptr
   %a = ptrtoint ptr %gp to i64
   %b = xor i64 7, %a
   %gp2 = inttoptr i64 %b to ptr
   store i16 0, ptr %gp2, align 2
```
This patch tries to infer the unchanged leading bit for the address
and let the target determine if it is safe to perform address space
cast for inttoptr instruction.

Commit:
[InferAS] Infer the address space of inttoptr

Currently the InferAddressSpaces would check if the bit value doesn't
change for <ptrtoint, inttoptr> address space cast. However the
condition is too strict. Sometime only the low bit address changes for
swizzling, and the address space is not changed. Take below code as
an example, we can tranform `%gp2 = inttoptr i64 %b to ptr` to
`%gp2 = inttoptr i64 %b to ptr addrspace(2)` and specify addrspace(2)
for the following store instruction.
```
   %gp = addrspacecast ptr addrspace(2) %sp to ptr
   %a = ptrtoint ptr %gp to i64
   %b = xor i64 7, %a
   %gp2 = inttoptr i64 %b to ptr
   store i16 0, ptr %gp2, align 2
```
This patch tries to infer the unchanged leading bit for the address
and let the target determine if it is safe to perform address space
cast for inttoptr instruction.

arsenm reviewed (changes requested):
This is more likely to be subtlety wrong than useful. I'd rather not even consider touching inttoptr

Artem-B commented:
I'm on the fence. 

On one hand, if we can prove that the bit changes keep the pointer value within the same AS, it would be useful to keep AS inference working.

On the other hand, NVIDIA does not give us anything specific about the AS values. We can empirically observe the pointers, but that's an implementation detail we can't really rely on. 

We can probably make a somewhat safe assumption that LSB changes within a "page" should keep the pointer within the same AS, but that's about it, and the patch appears to restrict the inference to roughly that. In principle it does look OK to me.

@arsenm if we guarantee that only "safe" LSB bits were changed, what could cause things to go wrong?
Being able to do limited arithmetic pointer manipulations is useful on NVIDIA GPUs these days, and it does require AS inference to work with it. Right now we often have to resort to inline asm.

arsenm commented:
> @arsenm if we guarantee that only "safe" LSB bits were changed, what could cause things to go wrong?

I'm more worried about provenance issues than relying on undocumented pointer properties

Commit:
Support more pattern of logic operation for the address computation.

LuoYuanke commented on llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:
```diff
@@ -369,15 +384,56 @@ getPointerOperands(const Value &V, const DataLayout &DL,
+
+  auto *Xor = dyn_cast<Operator>(I2P->getOperand(0));
+  if (!Xor || Xor->getOpcode() != Instruction::Xor)
+    return false;
```
`InstCombineAndOrXor.cpp` have 5000+ lines of code to reduce the logic operation. Unfortunately `ValueTracking` doesn't provide an API to compute the unchanged bits. I think we can start with some patterns for swizzling and improve the inferring with more and more patterns.

LuoYuanke replied:
Hi @Artem-B, Based on your idea, I draft a patch (https://github.com/llvm/llvm-project/pull/175406). The test case shows that with InstCombine we are able to get the 0 value for the ptr difference for the high bits of pointer. If we agree on this approach, I'll go on to finish the patch.

nikic commented:
Is it possible to replace the ptrtoint + arithmetic + inttoptr with ptradd(p, new_p - p)? This is the canonical way to change the value of the pointer while keeping the provenance (and in this context also address space). The pattern should get simplified to simple arithmetic in the backend.

Generally you want to avoid the emission of ptrtoint/inttoptr at all costs.

LuoYuanke commented:
Sorry, I am not quite experienced with Clang frontend. C++ developer may write the the code https://godbolt.org/z/aeda83jv8, and frontend has to emit ptrtoint/inttoptr instructions.

Artem-B commented:
> My concern is if this could be introducing a different class of provenance issues. 

Why would that be the case? AFAICT, there are two possible outcomes depending on whether we were able to find a source pointer with a known AS and that we were able to prove that the pointer remained in the same AS after the inttoptr. If we did, then we would insert ASC(generic->original AS), and possibly later propagate AS up/down the pointer uses. That should not complicate provenance any more that it may have been complicated already. We may have few/more ASCs in the IR, but the data dependency graph complexity will remain largely unchanged. If we failed to prove the necessary conditions, then we do not change anything.

There could be second-order issues like "do arithmetic ops on integer representation remain valid if we move the pointer from the generic AS to the specific AS?". In this particular case, I believe the answer is "yes" for NVPTX, provided that we did succeed at proving that the operations only affected in-AS-window LSBs. On NVIDIA GPUs, all specific AS are mapped 1:1 into their own window in a generic AS, so in-window LSB manipulations remain valid. This may not be the case for other targets. 

Overall, I do not think that this patch materially changes the status quo on how we handle int/ptr conversion and provenance tracking.

Commit:
[InferAS] revise the TTI interface

LuoYuanke commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1927,6 +1927,20 @@ class TargetTransformInfo {
+  ///   %gp2 = inttoptr i64 %b to ptr
+  ///   store i16 0, ptr %gp2, align 2
+  LLVM_ABI bool isSafeToCastIntPtrWithAS(unsigned AddrChangedLSB,
+                                         unsigned SrcAS, unsigned DstAS) const;
```
Revised. It seems we can not deduce the changed bit of pointer from computeKnownBitsAddrSpaceCast. computeKnownBitsAddrSpaceCast only return the known bits of the same ptr in fromAS and toAS. We need to know the bits change of two different pointers.

arsenm commented on llvm/include/llvm/Analysis/TargetTransformInfoImpl.h:
```diff
@@ -193,6 +193,11 @@ class TargetTransformInfoImplBase {
 
+  virtual APInt getAddrSpaceCastPreservedPtrMask(unsigned SrcAS,
+                                                 unsigned DstAS) const {
+    return APInt::getZero(DL.getPointerSizeInBits());
```
This should not use the default address space size (actually is this supposed to return the source width, the dest width, or is it assumed they are the same?)

LuoYuanke replied:
I hesitated for the interface for long time. My original thought is to unify the bit width to default address space, so that it is easy to compute mask with other APInt value without worrying about the different bit width.
How about return std::optional\<APInt\> for this interface? By default, it return std::nullopt. Target can return any bit width of the mask and the upper layer code should aware the bit width (truncate or extend the value) before compute mask with other APInt value.

LuoYuanke commented on llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:
```diff
@@ -375,15 +390,81 @@ getPointerOperands(const Value &V, const DataLayout &DL,
+    return ~Known.Zero;
+  case Instruction::And:
+    return ~Known.One;
+  default:
```
For example "NewPtr = 7 or OldPtr", we need to know only lower 3 bits changes by comparing NewPtr with OldPtr. Compiler doesn't know the bits of OldPtr. Compiler can deduce the lower 3 bits is 1 for NewPtr, but it doesn't know the other bits of the NewPtr. To validate it is legal to perform address space cast, we need to know only the lower 3 bits changes which depends on operator `or` and operand `7`.

Commit:
Address Matt's comments

LuoYuanke commented on llvm/lib/Target/NVPTX/NVPTXTargetTransformInfo.h:
```diff
@@ -180,6 +180,19 @@ class NVPTXTTIImpl final : public BasicTTIImplBase<NVPTXTTIImpl> {
+  APInt getAddrSpaceCastPreservedPtrMask(unsigned SrcAS,
+                                         unsigned DstAS) const override {
+    if (SrcAS != llvm::ADDRESS_SPACE_GENERIC)
+      return APInt::getZero(DL.getPointerSizeInBits());
```
Revised

LuoYuanke commented on llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:
```diff
@@ -375,15 +390,81 @@ getPointerOperands(const Value &V, const DataLayout &DL,
+    return false;
+
+  auto *LHS = LogicOP->getOperand(0);
+  auto *Mask = LogicOP->getOperand(1);
```
Revised.

LuoYuanke commented on llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:
```diff
@@ -375,15 +390,81 @@ getPointerOperands(const Value &V, const DataLayout &DL,
+static APInt computeMaxChangedPtrBits(const Operator *Op, const Value *Mask,
+                                      const DataLayout &DL, AssumptionCache *AC,
+                                      const DominatorTree *DT) {
+  KnownBits Known = computeKnownBits(Mask, DL, AC, nullptr, DT);
```
We may support ConstantExprs as well. What do you think?

Commit:
Address comments.

Commit:
Revert "Address comments."

This reverts commit 26e6e78eedecb10f2191edd58c9033fed1408a15.
The logical operator pattern match doesn't work for constant
expression, so revert the patch

Commit:
[InferAS] Address the comments.

1. revert the inferface getAddrSpaceCastPreservedPtrMask to return APInt
2. emit new addrspacecast instead of emit inttoptr with new address space.

Commit:
[InferAS] call BaseT method when it has nothing to do.

Commit:
[InferAS] return early when PreservedPtrMask is zero.

arsenm commented on llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:
```diff
@@ -375,15 +390,81 @@ getPointerOperands(const Value &V, const DataLayout &DL,
+    std::swap(LHS, Mask);
+  P2I = dyn_cast<Operator>(LHS);
+  if (!P2I || P2I->getOpcode() != Instruction::PtrToInt)
+    return false;
```
That's fixable, but I would also consider that not important enough to warrant not using match. If you really want to handle the constant expression case, extend the support to m_c_BitwiseLogic later

LuoYuanke replied:
Revised to pattern match.

Commit:
[InferAS] Address the comments.

1. Use pattern match to detect inttoptr
2. Rename isSafeToCastPtrIntPair to isSafeToCastIntToPtrAddrSpace
3. Refine comments

LuoYuanke commented on llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:
```diff
@@ -375,15 +391,90 @@ getPointerOperands(const Value &V, const DataLayout &DL,
+  }
+}
+
+bool InferAddressSpacesImpl::isSafeToCastPtrIntPair(
```
Renamed

Commit:
[InferAS] Fix clang format issues

Commit:
Add getIntToPtrPointerOperand() API

shiltian commented on llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp:
Oh yeah I see the issue. Well, then I'd say then it is probably a good idea to instead of lazily populating the map inside `getIntToPtrPointerOperand`, add a dedicated non-const pre-population step at the start of `run()`. In this way, it is no need for the `mutable`.

Commit:
Collect inttoptr original pointer at the begin of run().

shiltian reviewed (approved):
I think this PR is in a fairly good shape now. Maybe hold for a day or two to see if @arsenm has more comments.

jayfoad commented:
Would it be OK to revert this until we have a definition for the hook that makes sense?

LuoYuanke commented:
> Would it be OK to revert this until we have a definition for the hook that makes sense?

I prefer not. It is helpful to optimize swizzle cases. For AMDGPU it is not enabled, so I think this patch doesn't affect AMDGPU target.

jayfoad commented:
> > Would it be OK to revert this until we have a definition for the hook that makes sense?
> 
> I prefer not. It is helpful to optimize swizzle cases. For AMDGPU it is not enabled, so I think this patch doesn't affect AMDGPU target.

I am not saying that it breaks AMDGPU. I am saying that the target hook seems to make no sense (in particular it seems impossible to implement it properly for AMDGPU) and should never have been merged in the first place.

arsenm commented:
> > Would it be OK to revert this until we have a definition for the hook that makes sense?
> 
> I prefer not. It is helpful to optimize swizzle cases. For AMDGPU it is not enabled, so I think this patch doesn't affect AMDGPU target.

Ping, just because it's not implemented for AMDGPU doesn't mean it's not broken

LuoYuanke commented:
I'd like to understand what's the issue in current PR. Is there anything to improve instead of revert? 
Consider this existing pattern that the pass already handles:
```
  %p = addrspacecast ptr addrspace(3) %shared to ptr
  %q = getelementptr i8, ptr %p, i64 %big_offset
  store i32 0, ptr %q
```
 The pass will infer AS 3 for %q and rewrite the store to ptr addrspace(3), without checking whether %big_offset pushes the pointer out of the shared memory window. A sufficiently large offset could produce a pointer that is no longer a valid shared memory address, yet the pass infers the address space anyway.

Now compare with the inttoptr case this PR handles:
```
  %p = addrspacecast ptr addrspace(3) %sp to ptr
  %a = ptrtoint ptr %p to i64
  %b = xor i64 %a, 7
  %q = inttoptr i64 %b to ptr
  store i32 0, ptr %q
```
  Here, KnownBits proves that the high 61 bits of the address are unchanged. This is actually a stronger guarantee than the GEP case, where the pass does no bounds checking at all — it just assumes the pointer arithmetic stays within the same AS window.
 If the GEP-based inference is acceptable (and it has been in production for years), I don't see a strong reason to reject the inttopr case. The inttoptr path is conservative, and the TTI hook defaults to false.
